compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/bir/writer/BIRInstructionWriter.java

```diff
@@ -0,0 +1,606 @@
+            buf.writeByte(0);
+        }
+        addCpAndWriteString(birCall.thenBB.id.value);
+    }
+
+    public void visit(BIRTerminator.AsyncCall birAsyncCall) {
+        writePosition(birAsyncCall.pos);
+        buf.writeByte(birAsyncCall.kind.getValue());
+        PackageID calleePkg = birAsyncCall.calleePkg;
+        int pkgIndex = addPkgCPEntry(calleePkg);
+        buf.writeBoolean(birAsyncCall.isVirtual);
+        buf.writeInt(pkgIndex);
+        buf.writeInt(addStringCPEntry(birAsyncCall.name.getValue()));
+        buf.writeInt(birAsyncCall.args.size());
+        for (BIROperand arg : birAsyncCall.args) {
+            arg.accept(this);
+        }
+        if (birAsyncCall.lhsOp != null) {
+            buf.writeByte(1);
+            birAsyncCall.lhsOp.accept(this);
+        } else {
+            buf.writeByte(0);
+        }
+
+        binaryWriter.writeAnnotAttachments(buf, this, birAsyncCall.annotAttachments);
+        addCpAndWriteString(birAsyncCall.thenBB.id.value);
+    }
+
+    public void visit(BIRTerminator.FPCall fpCall) {
+        writePosition(fpCall.pos);
+        buf.writeByte(fpCall.kind.getValue());
+        fpCall.fp.accept(this);
+        buf.writeInt(fpCall.args.size());
+        for (BIROperand arg : fpCall.args) {
+            arg.accept(this);
+        }
+        if (fpCall.lhsOp != null) {
+            buf.writeByte(1);
+            fpCall.lhsOp.accept(this);
+        } else {
+            buf.writeByte(0);
+        }
+        buf.writeBoolean(fpCall.isAsync);
+        addCpAndWriteString(fpCall.thenBB.id.value);
+    }
+
+    public void visit(BIRNonTerminator.BinaryOp birBinaryOp) {
+        writePosition(birBinaryOp.pos);
+        buf.writeByte(birBinaryOp.kind.getValue());
+        birBinaryOp.rhsOp1.accept(this);
+        birBinaryOp.rhsOp2.accept(this);
+        birBinaryOp.lhsOp.accept(this);
+    }
+
+    public void visit(BIRNonTerminator.UnaryOP birUnaryOp) {
+        writePosition(birUnaryOp.pos);
+        buf.writeByte(birUnaryOp.kind.getValue());
+        birUnaryOp.rhsOp.accept(this);
+        birUnaryOp.lhsOp.accept(this);
+    }
+
+    public void visit(BIRNonTerminator.ConstantLoad birConstantLoad) {
+        writePosition(birConstantLoad.pos);
+        buf.writeByte(birConstantLoad.kind.getValue());
+        writeType(birConstantLoad.type);
+        birConstantLoad.lhsOp.accept(this);
+
+        BType type = birConstantLoad.type;
+        switch (type.tag) {
+            case TypeTags.INT:
+            case TypeTags.SIGNED32_INT:
+            case TypeTags.SIGNED16_INT:
+            case TypeTags.SIGNED8_INT:
+            case TypeTags.UNSIGNED32_INT:
+            case TypeTags.UNSIGNED16_INT:
+            case TypeTags.UNSIGNED8_INT:
+                buf.writeInt(cp.addCPEntry(new IntegerCPEntry((Long) birConstantLoad.value)));
+                break;
+            case TypeTags.BYTE:
+                // TODO: birConstantLoad.value should return an Integer. This is a temporary fix
```

Comment:
Shall we create an issue about why it should return an integer to make sure it is fixed?